luminoso_api/v5_upload.py

```diff
@@ -0,0 +1,200 @@
+import argparse
+import json
+import time
+import sys
+from urllib.parse import urlparse
+from itertools import islice, chain
+from tqdm import tqdm
+
+from .v5_cli import connect_with_token_args
+from .v5_client import LuminosoClient
+from .errors import LuminosoServerError
+from .v5_constants import URL_BASE
+
+
+DESCRIPTION = 'Create a Luminoso project from documents in a file.'
+
+# These fields (and only these fields) must exist on every uploaded document.
+UPLOAD_FIELDS = ['title', 'text', 'metadata']
+BATCH_SIZE = 1000
+
+# http://code.activestate.com/recipes/303279-getting-items-in-batches/
+# Updated for Python 3.5+ by catching StopIteration
+def _batches(iterable, size):
+    """
+    Take an iterator and yield its contents in groups of `size` items.
+    """
+    sourceiter = iter(iterable)
+    while True:
+        try:
+            batchiter = islice(sourceiter, size)
+            yield chain([next(batchiter)], batchiter)
+        except StopIteration:
+            return
+
+
+def _simplify_doc(doc):
+    """
+    Limit a document to just the three fields we should upload.
+    """
+    # Mutate a copy of the document to fill in missing fields
+    doc = dict(doc)
+    simplified = {}
+    if 'text' not in doc:
+        raise ValueError("The document {!r} has no text field".format(doc))
+    if 'metadata' not in doc:
+        doc['metadata'] = []
+    if 'title' not in doc:
+        doc['title'] = ""
+
+    for field in UPLOAD_FIELDS:
+        simplified[field] = doc[field]
```

Comment:
This entire function actually seems weird to me--we're taking the supplied document, copying it, and modifying it; and then reading fields out of it into another dictionary.  Why not just build the other dictionary from the start?  That is,

```python
def _simplify_doc(doc):
    if 'text' not in doc:
        raise ValueError("The document {!r} has no text field".format(doc))
    return {'text': doc['text'], 'metadata': doc.get('metadata', []), 'title': doc.get('title', '')}
```
This admittedly has the disadvantage of hard-coding the three relevant fields into the function rather than using UPLOAD_FIELDS.  On the other hand, that's already the case: if something new is added to UPLOAD_FIELDS without another `if <that_field> not in doc:` block added, this function will fail with a KeyError anyway.